compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs

```diff
@@ -1955,6 +1964,410 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                        }
+                        hir::def::DefKind::Variant => {
+                            let variant_def_id = *struct_def_id;
+                            let struct_def_id = self.tcx.parent(variant_def_id);
+
+                            if impl_self_ty_path.did() != struct_def_id {
+                                // If the struct is not the same as `Self`, we cannot refine.
+                                return Err(expr);
+                            }
+
+                            let struct_def_generics: &ty::Generics =
+                                self.tcx.generics_of(struct_def_id);
+
+                            let struct_variant_def: Option<&ty::VariantDef> = impl_self_ty_path
+                                .variants()
+                                .iter()
+                                .find(|variant: &&ty::VariantDef| variant.def_id == variant_def_id);
+
+                            let Some(struct_variant_def) = struct_variant_def else {
+                                // Failed to find matching variant.
+                                return Err(expr);
+                            };
+
+                            // Use the sole variant definition as our variant from now on:
+                            (struct_variant_def, struct_def_generics)
+                        }
+                        _ => {
+                            return Err(expr);
+                        }
+                    };
+
+                // Only retain the generics which are relevant (according to the `impl`).
+                let struct_def_generics: Vec<&ty::GenericParamDef> = relevant_ty_args_indices
+                    .into_iter()
+                    .filter_map(|index| struct_def_generics.params.get(index))
+                    .collect();
+
+                // The struct being constructed is the same as the struct in the impl.
+
+                let blameable_field_defs: Vec<&ty::FieldDef> = struct_variant_def
+                    .fields
+                    .iter()
+                    .filter(|field_def| {
+                        let field_type: Ty<'tcx> = self.tcx.type_of(field_def.did);
+
+                        // Only retain fields which mention the blameable generics.
+                        struct_def_generics
+                            .iter()
+                            .any(|param| find_param_def_in_ty_walker(field_type.walk(), param))
+                    })
+                    .collect();
+
+                if blameable_field_defs.len() != 1 {
+                    // We must have a single blameable field, in order to be able to blame it.
+                    return Err(expr);
+                }
+
+                let blameable_field_def: &ty::FieldDef = blameable_field_defs[0];
+
+                for field in struct_fields.iter() {
+                    if field.ident.as_str() == blameable_field_def.ident(self.tcx).as_str() {
```

Comment:
I think calling `.ident(..).normalize_to_macros_2_0()` then comparing the `Ident`s themselves is technically more correct -- see https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_hir_typeck/expr.rs.html#1567 for example.

If you're not convinced that that's correct, then you should at least be comparing `ident.name` as [`Symbol`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/symbol/struct.Symbol.html)

Reply:
I'm 100% sure that the way I'm doing it right now is wrong 🙃. I just had no idea what the right way looked like. This seems much better. I'll get this in my next pass.

Reply:
I was able to reuse `point_at_field_if_possible`, eliminating this logic